### PR TITLE
Avoid `Expr` copies `OptimizeProjection`, 12% faster planning, encapsulate indicies

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/10209

## Rationale for this change

This is part 1 of improving the performance of `OptimizeProjection` (by less copying) in https://github.com/apache/datafusion/issues/10209.

While trying to rework how `OptimizeProjection works`, I found that the population
of child indices pretty much required calling `LogicalPlan::expressions()` which results in an owed copy of all the plans `Expr`s

Removing these copies themselves improves the situation (and I think may make the code easier to reasona about), and sets us up to stop copying the `LogicalPlan` nodes as a follow on PR

## What changes are included in this PR?

This PR encapsulates that logic for managing required indices into a new struct `RequiredIndicies` which avoids some Expr copies (and sets up the rewrite to avoid `LogicalPlan` copies)

## Are these changes tested?
Covered by existing tests

## Are there any user-facing changes?
no functional changes

performance benchmarks show an overall 14% improvements for tpch_all and tpcds_all, which is pretty neat

<details><summary>Details</summary>
<p>

```
++ critcmp main optimize_proj_pt1
group                                         main                                   optimize_proj_pt1
-----                                         ----                                   -----------------
logical_aggregate_with_join                   1.02  1210.2±63.66µs        ? ?/sec    1.00  1188.2±32.04µs        ? ?/sec
logical_plan_tpcds_all                        1.02    158.0±1.26ms        ? ?/sec    1.00    154.3±1.56ms        ? ?/sec
logical_plan_tpch_all                         1.05     17.0±0.20ms        ? ?/sec    1.00     16.3±0.18ms        ? ?/sec
logical_select_all_from_1000                  1.04     19.2±0.16ms        ? ?/sec    1.00     18.5±0.09ms        ? ?/sec
logical_select_one_from_700                   1.01   792.6±42.68µs        ? ?/sec    1.00    787.7±7.31µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.01   743.5±10.96µs        ? ?/sec    1.00   732.9±28.34µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.01    729.8±8.98µs        ? ?/sec    1.00    720.9±9.40µs        ? ?/sec
physical_plan_tpcds_all                       1.14  1500.3±10.49ms        ? ?/sec    1.00  1311.0±13.83ms        ? ?/sec
physical_plan_tpch_all                        1.14     99.9±0.72ms        ? ?/sec    1.00     87.9±0.39ms        ? ?/sec
physical_plan_tpch_q1                         1.14      5.4±0.02ms        ? ?/sec    1.00      4.7±0.02ms        ? ?/sec
physical_plan_tpch_q10                        1.10      4.6±0.02ms        ? ?/sec    1.00      4.2±0.02ms        ? ?/sec
physical_plan_tpch_q11                        1.08      4.0±0.02ms        ? ?/sec    1.00      3.7±0.02ms        ? ?/sec
physical_plan_tpch_q12                        1.11      3.3±0.01ms        ? ?/sec    1.00      3.0±0.01ms        ? ?/sec
physical_plan_tpch_q13                        1.08      2.2±0.01ms        ? ?/sec    1.00      2.1±0.01ms        ? ?/sec
physical_plan_tpch_q14                        1.11      2.9±0.02ms        ? ?/sec    1.00      2.6±0.03ms        ? ?/sec
physical_plan_tpch_q16                        1.12      4.0±0.03ms        ? ?/sec    1.00      3.6±0.02ms        ? ?/sec
physical_plan_tpch_q17                        1.12      3.8±0.03ms        ? ?/sec    1.00      3.4±0.02ms        ? ?/sec
physical_plan_tpch_q18                        1.11      4.2±0.04ms        ? ?/sec    1.00      3.8±0.02ms        ? ?/sec
physical_plan_tpch_q19                        1.26      8.0±0.09ms        ? ?/sec    1.00      6.3±0.04ms        ? ?/sec
physical_plan_tpch_q2                         1.11      8.5±0.07ms        ? ?/sec    1.00      7.6±0.03ms        ? ?/sec
physical_plan_tpch_q20                        1.12      4.9±0.04ms        ? ?/sec    1.00      4.4±0.02ms        ? ?/sec
physical_plan_tpch_q21                        1.14      6.8±0.07ms        ? ?/sec    1.00      6.0±0.03ms        ? ?/sec
physical_plan_tpch_q22                        1.11      3.6±0.04ms        ? ?/sec    1.00      3.3±0.03ms        ? ?/sec
physical_plan_tpch_q3                         1.09      3.3±0.01ms        ? ?/sec    1.00      3.0±0.02ms        ? ?/sec
physical_plan_tpch_q4                         1.08      2.4±0.01ms        ? ?/sec    1.00      2.2±0.01ms        ? ?/sec
physical_plan_tpch_q5                         1.11      4.8±0.02ms        ? ?/sec    1.00      4.3±0.02ms        ? ?/sec
physical_plan_tpch_q6                         1.11  1735.8±11.71µs        ? ?/sec    1.00   1557.8±9.94µs        ? ?/sec
physical_plan_tpch_q7                         1.13      6.2±0.04ms        ? ?/sec    1.00      5.5±0.02ms        ? ?/sec
physical_plan_tpch_q8                         1.12      8.0±0.07ms        ? ?/sec    1.00      7.1±0.02ms        ? ?/sec
physical_plan_tpch_q9                         1.12      6.0±0.04ms        ? ?/sec    1.00      5.4±0.02ms        ? ?/sec
physical_select_all_from_1000                 1.49     90.9±0.30ms        ? ?/sec    1.00     60.9±0.20ms        ? ?/sec
physical_select_one_from_700                  1.08      3.7±0.05ms        ? ?/sec    1.00      3.4±0.02ms        ? ?/sec
```

</p>
</details>

### CODE REVIEW

File: datafusion/common/src/dfschema.rs [hunk 1]
```diff
@@ -347,7 +347,20 @@ impl DFSchema {
...
     }
 
-    /// Find the index of the column with the given qualifier and name
+    /// Find the index of the column with the given qualifier and name,
+    /// returning `None` if not found
+    ///
+    /// See [Self::index_of_column] for a version that returns an error if the
+    /// column is not found
+    pub fn maybe_index_of_column(&self, col: &Column) -> Option<usize> {
+        self.index_of_column_by_name(col.relation.as_ref(), &col.name)
```
Comment (alamb) [inline datafusion/common/src/dfschema.rs:356]: This hung me up for a while -- the original code uses `flat_map` which discards the `Err` if the column is not found, was not obvious to me

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 2]
```diff
@@ -105,13 +106,9 @@ impl OptimizerRule for OptimizeProjections {
 fn optimize_projections(
     plan: &LogicalPlan,
     config: &dyn OptimizerConfig,
-    indices: &[usize],
+    indices: RequiredIndicies,
 ) -> Result<Option<LogicalPlan>> {
-    // `child_required_indices` stores
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:110]: this documentation moved to `RequiredIndices`

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 3]
```diff
@@ -339,31 +318,35 @@ fn optimize_projections(
             let left_len = join.left.schema().fields().len();
             let (left_req_indices, right_req_indices) =
                 split_join_requirements(left_len, indices, &join.join_type);
-            let exprs = plan.expressions();
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:342]: This call copies all expressions in the plan

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 4]
```diff
@@ -123,12 +120,13 @@ fn optimize_projections(
             // that appear in this plan's expressions to its child. All these
             // operators benefit from "small" inputs, so the projection_beneficial
             // flag is `true`.
-            let exprs = plan.expressions();
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:126]: Likewise, this was  also copying all expressions

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 5]
```diff
@@ -137,13 +135,9 @@ fn optimize_projections(
             // that appear in this plan's expressions to its child. These operators
             // do not benefit from "small" inputs, so the projection_beneficial
             // flag is `false`.
-            let exprs = plan.expressions();
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:140]: And another copy removed

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 6]
```diff
@@ -178,16 +170,12 @@ fn optimize_projections(
                 Make sure `.necessary_children_exprs` implementation of the `UserDefinedLogicalNode` is \
                 consistent with actual children length for the node.");
             }
-            // Expressions used by node.
-            let exprs = plan.expressions();
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:182]: Another copy

File: datafusion/optimizer/src/optimize_projections/mod.rs [hunk 7]
```diff
@@ -408,26 +392,6 @@ fn optimize_projections(
     }
 }
 
-/// This function applies the given function `f` to the projection indices
```
Comment (alamb) [inline datafusion/optimizer/src/optimize_projections/mod.rs:411]: All of those code was moved into `RequiredIndices` as methods both to encapsulate the logic a bit more as well as to allow creation without copying

File: datafusion/common/src/dfschema.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/common/src/dfschema.rs:356]: This is much more clear, Thanks!

**Review by mustafasrepo — APPROVED** [review]
Thanks @alamb. This PR is LGTM!. As well as improving performance, it improves maintainability also.

**Comment by alamb** [comment]
Thanks for the reiew @mustafasrepo  -- I am slowly working through how to refactor this pass to avoid copies. I think a few more PRs and I'll have it

**Pull request merged**